### PR TITLE
fix: Use actual replayed tx's  gas and pubdata prices to replay tx.

### PR DESCRIPTION
**Pull request by AnastasiiaVashchuk**
# What :computer: 
+ bump `zksync era` dependencies version
+ new get_block_gas_details in `fork.rs` to extract l1 batch gas info.
+ modify `create_l1_batch_env` to construct `BatchFeeInput` from values taken from l1 batch details, if memory node is running in fork mode.

# Why :hand:
Context: successful txs from fork fails while replaying them with node. 
This is caused by wrong gas/fee values stored in `fee_input_provider` that are used to calculate tx cost.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -12,16 +12,15 @@ publish = false                                           # We don't want to pub
...
-zksync_basic_types = { git = "https://github.com/matter-labs/zksync-era.git", rev = "e10bbdd1e863962552f37e768ae6af649353e4ea" }
-zksync_node_fee_model = { git = "https://github.com/matter-labs/zksync-era.git", rev = "e10bbdd1e863962552f37e768ae6af649353e4ea" }
-multivm = { git = "https://github.com/matter-labs/zksync-era.git", rev = "e10bbdd1e863962552f37e768ae6af649353e4ea" }
-zksync_contracts = { git = "https://github.com/matter-labs/zksync-era.git", rev = "e10bbdd1e863962552f37e768ae6af649353e4ea" }
-zksync_types = { git = "https://github.com/matter-labs/zksync-era.git", rev = "e10bbdd1e863962552f37e768ae6af649353e4ea" }
-zksync_utils = { git = "https://github.com/matter-labs/zksync-era.git", rev = "e10bbdd1e863962552f37e768ae6af649353e4ea" }
-zksync_state = { git = "https://github.com/matter-labs/zksync-era.git", rev = "e10bbdd1e863962552f37e768ae6af649353e4ea" }
-zksync_web3_decl = { git = "https://github.com/matter-labs/zksync-era.git", rev = "e10bbdd1e863962552f37e768ae6af649353e4ea", features = [
+zksync_basic_types = { git = "https://github.com/matter-labs/zksync-era.git", rev = "0d51cd6f3e65eef1bda981fe96f3026d8e12156d" }
```
Comment (AnastasiiaVashchuk) [inline Cargo.toml:15]: in this commit there was added `fair_pubdata_price` value, that is needed for replaying transaction.
more details here: 
https://github.com/matter-labs/zksync-era/pull/2244

File: Cargo.toml [hunk 2]
```diff
@@ -12,16 +12,15 @@ publish = false                                           # We don't want to pub
...
+zksync_node_fee_model = { git = "https://github.com/matter-labs/zksync-era.git", rev = "0d51cd6f3e65eef1bda981fe96f3026d8e12156d" }
+multivm = { git = "https://github.com/matter-labs/zksync-era.git", rev = "0d51cd6f3e65eef1bda981fe96f3026d8e12156d" }
+zksync_contracts = { git = "https://github.com/matter-labs/zksync-era.git", rev = "0d51cd6f3e65eef1bda981fe96f3026d8e12156d" }
+zksync_types = { git = "https://github.com/matter-labs/zksync-era.git", rev = "0d51cd6f3e65eef1bda981fe96f3026d8e12156d" }
+zksync_utils = { git = "https://github.com/matter-labs/zksync-era.git", rev = "0d51cd6f3e65eef1bda981fe96f3026d8e12156d" }
+zksync_state = { git = "https://github.com/matter-labs/zksync-era.git", rev = "0d51cd6f3e65eef1bda981fe96f3026d8e12156d" }
+zksync_web3_decl = { git = "https://github.com/matter-labs/zksync-era.git", rev = "0d51cd6f3e65eef1bda981fe96f3026d8e12156d", features = [
     "server",
-    "client",
```
Comment (AnastasiiaVashchuk) [inline Cargo.toml:24]: zksync_web3_decl does not support `client` feature on this commit

File: src/node/eth.rs [hunk 3]
```diff
@@ -1390,14 +1390,11 @@ impl<S: ForkSource + std::fmt::Debug + Clone + Send + Sync + 'static> EthTestNod
         &self,
         tx: zksync_types::transaction_request::CallRequest,
     ) -> jsonrpc_core::BoxFuture<jsonrpc_core::Result<zksync_basic_types::H256>> {
-        let (chain_id, l1_gas_price) = match self.get_inner().read() {
-            Ok(reader) => (
-                reader.fork_storage.chain_id,
-                reader.fee_input_provider.l1_gas_price,
-            ),
+        let l1_gas_price = match self.get_inner().read() {
+            Ok(reader) => reader.fee_input_provider.l1_gas_price,
```
Comment (AnastasiiaVashchuk) [inline src/node/eth.rs:1394]: changes in this file are caused by bumping version if zksync_ dependencies

File: src/utils.rs [hunk 4]
```diff
@@ -265,7 +265,9 @@ pub fn report_into_jsrpc_error(error: eyre::Report) -> Error {
 pub fn into_jsrpc_error(err: Web3Error) -> Error {
     Error {
         code: match err {
-            Web3Error::InternalError(_) | Web3Error::NotImplemented => ErrorCode::InternalError,
+            Web3Error::InternalError(_) | Web3Error::MethodNotImplemented => {
+                ErrorCode::InternalError
+            }
             Web3Error::NoBlock
```
Comment (AnastasiiaVashchuk) [inline src/utils.rs:271]: changes are caused by bumping the version of zksync_ dependencies

File: src/node/eth.rs [hunk 5]
```diff
@@ -1429,18 +1426,26 @@ impl<S: ForkSource + std::fmt::Debug + Clone + Send + Sync + 'static> EthTestNod
...
+                tracing::error!("Transaction request serialization error: {}", e);
+                return futures::future::err(into_jsrpc_error(Web3Error::SerializationError(e)))
+                    .boxed();
+            }
+        };
+        let bytes = match tx_req.get_signed_bytes(&PackedEthSignature::from_rsv(
+            &H256::default(),
+            &H256::default(),
+            27,
+        )) {
```
Comment (AnastasiiaVashchuk) [inline src/node/eth.rs:1441]: changes are caused by bumping the version of zksync_ dependencies

File: src/fork.rs [hunk 6]
```diff
@@ -644,6 +644,26 @@ impl ForkDetails {
             replay_tx, miniblock
         );
     }
+
+    pub fn get_block_gas_details(&self, miniblock: u32) -> Option<(u64, u64, u64)> {
```
Comment (mm-zk) [inline src/fork.rs:648]: please add comment explaining the 3 things that are being returned.

File: src/fork.rs [hunk 7]
```diff
@@ -644,6 +644,26 @@ impl ForkDetails {
...
+        let res_opt_block_details = self.fork_source.get_block_details(L2BlockNumber(miniblock));
+        match res_opt_block_details {
+            Ok(opt_block_details) => opt_block_details.map(|block_details| {
+                (
+                    block_details.base.l1_gas_price,
+                    block_details.base.l2_fair_gas_price,
+                    block_details
+                        .base
+                        .fair_pubdata_price
+                        .expect("fair pubdata price is not present in block details"),
```
Comment (mm-zk) [inline src/fork.rs:658]: mention the miniblock number inside the expect

**Review by mm-zk — DISMISSED** [review]

**Review by mm-zk — APPROVED** [review]

**Pull request merged**